# Bioinformatics Algorithms

## Overview
This repository contains my personal implementations of core algorithms from Bioinformatics Algorithms: An Active Learning Approach by Phillip Compeau & Pavel Pevzner, exploring computational techniques widely used in bioinformatics. Each module focuses on understanding, implementing, and testing algorithms from scratch, without relying on high-level libraries. It serves both as a learning record and a reference for key algorithmic concepts in computational biology.

---

## Table of Contents:
- [Overview](#bioinformatics-algorithms)
- [Covered Topics](#covered-topics)
- [Project Structure](#project-structure)
- [Installation](#installation)
- [Usage](#usage)
- [Contributing](#contributing)
- [References](#references)

---

## Covered Topics
- **Origin Finding** – detecting replication origins in bacterial genomes  
- **Motif Search** – identifying common motifs across DNA sequences  
- **Genome Sequencing** – assembling genomes from reads or paired reads
- **Peptide Sequencing** – reconstructing cyclic peptides from mass spectrometry  
- **Sequence Alignment** – aligning DNA, RNA, and protein sequences  
- **Genome Rearrangement** – finding minimum 2-break operations between two signed genome permutations
- **Phylogenetic Analysis** – building evolutionary trees  
- **Clustering Analysis** – grouping biological data based on similarity  
- **Combinatorial Pattern Matching** – efficient pattern search algorithms  
- **HMM Classification Analysis** – sequence classification using profile hidden Markov model matrices
- **Peptide Vector Search** – sequences linear peptides from vector spectra and searches for database matches.

---

## Project Structure

```
Bioinformatics-Algorithms/
├── 01_origin_finder/
│   ├── main.py
│   ├── origin_finder.py
│   ├── test.txt
│   └── results.txt
├── 02_motif_search/
│   ├── main.py
│   ├── motif_search.py
│   ├── test.txt
│   └── results.txt
├── 03_genome_sequencing/
│   ├── main.py
│   ├── genome_sequencing.py
│   ├── reads_and_pairs.txt
│   ├── test.txt
│   └── results.txt
├── 04_peptide_sequencing/
│   ├── main.py
│   ├── peptide_sequencing.py
│   ├── test.txt
│   └── results.txt
├── 05_sequence_alignment/
│   ├── main.py
│   ├── sequence_alignment.py
│   ├── local_test.txt
│   ├── global_test.txt
│   ├── fitting_test.txt
│   ├── msa_test.txt
│   └── results.txt
├── 06_genome_rearrangement/
│   ├── main.py
│   ├── genome_rearrangement.py
│   ├── 2b_test1.txt
│   ├── 2b_test2.txt
│   ├── bp_test1.txt
│   ├── bp_test2.txt
│   └── results.txt
├── 07_phylogenetic_analysis/
│   ├── main.py
│   ├── phylogenetic_analysis.py
│   ├── test.txt
│   └── results.txt
├── 08_clustering_analysis/
│   ├── main.py
│   ├── clustering_analysis.py
│   ├── test.txt
│   └── results.txt
├── 09_combinatorial_pattern_matching/
│   ├── main.py
│   ├── pattern_matching.py
│   ├── test1.txt
│   ├── test2.txt
│   ├── reads.txt
│   └── results.txt
├── 10_hmm_classification_analysis/
│   ├── main.py
│   ├── hmm_classification.py
│   ├── test1.txt
│   ├── test2.txt
│   ├── hmm_matrices.txt
│   └── results.txt
├── 11_peptide_vector_search/
│   ├── main.py
│   ├── peptide_search.py
│   ├── test1.txt
│   ├── test2.txt
│   └── results.txt
└── README.md
```

## Installation

## Usage

## Contributing

## References

